DSpace/dspace-angular
# Listable-object decorator refactor

## References
Fixes #1911

## Description
This PR refactors listable-object decorator to use a new system for determining the relevancy between matching components and returning the most relevant one for the inputs provided.
Previously, the code would ignore some inputs the user provided in certain cases (see issue), this should fix that.

## Instructions for Reviewers
Changes made:
- listable-object decorator has a new internal class called `MatchRelevancy`. This class stores a match found within the map and information about how it was found. It stores the first "level" a default value was used and the "relevancy" representing the amount of non-default values were used to find this match.
- Using this class, getListableObjectComponent can determine which of the found matches is the most relevant one.
- Retrieving a match wrapped within a MatchRelevancy is done through a new getMatch() method, which works roughly the same way it used to, but touches on the issues the previous code had, whilst simultaneously making it open for change (e.g. easy to add a new parameter to our decorator)
- All of this is well documented within the code, so please check out the JSDocs added in this PR for more details and examples

How to test:
- The original test cases still succeed, pointing towards me not breaking any existing functionality
- New test cases have been added that cover my edge-cases from the issue and more
- These new tests fail when the code is reverted to its original state
- The UI should be unaffected if you visit different list views, item pages, etc.
- Optionally, you can take the example from the issue and confirm the right component is resolved after my changes

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

    * Include tests for different user types (if behavior differs), including: (1) Anonymous user, (2) Logged in user (non-admin), and (3) Administrator.
    * Include tests for error scenarios, e.g. when errors/warnings should appear (or buttons should be disabled).
    * For bug fixes, include a test that reproduces the bug and proves it is fixed. For clarity, it may be useful to provide the test in a separate commit from the bug fix.
